src/other/utils.jl

```diff
@@ -131,6 +131,115 @@ end
 
+
+const SINGLETHREADING = Threads.Atomic{Bool}(false)
+const SINGLETHREADING_DEPTH = Threads.Atomic{Int}(0)
+
+ismultithreaded() = SINGLETHREADING_DEPTH[] == 0
+
+"""
+    DataFrames.singlethreaded(f)
+
+Run function `f` while disabling multithreading in all DataFrames.jl operations.
+This is useful in particular to run functions which are not thread-safe, or when
+distribution of work across threads is managed separately.
+
+*See also*: [`DataFrames.setmultithreading`](@ref) to disable multithreading globally
+
+!!! note
+
+    This function is considered as experimental and may change or be removed once
+    a cross-package mechanism for multithreading configuration is developed.
+
+    Currently, it disables multithreading for any DataFrames.jl
+    operations which may be run while `f` is running (e.g. if tasks using data
+    frames have been spawned on multiple threads).
+    This may change in the future.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(x=repeat(1:5, inner=2), y=1:10);
+
+julia> gd = groupby(df, :x);
+
+julia> counter = 0;
+
+julia> f(x) = (sleep(0.1); global counter += 1); # Thread-unsafe function
+
+julia> DataFrames.singlethreaded() do
+           combine(gd, :y => f)
+       end
+5×2 DataFrame
+ Row │ x      y_f   
+     │ Int64  Int64 
+─────┼──────────────
+   1 │     1      1
+   2 │     2      2
+   3 │     3      3
+   4 │     4      4
+   5 │     5      5
+```
+"""
+function singlethreaded(f)
+    Threads.atomic_add!(SINGLETHREADING_DEPTH, 1)
+    try
+        return f()
+    finally
+        Threads.atomic_sub!(SINGLETHREADING_DEPTH, 1)
+    end
+end
+
+"""
+    DataFrames.setmultithreading(enable::Bool)
+
+Enable or disable multithreading permanently in all DataFrames.jl operations.
+This is useful in particular to run functions which are not thread-safe, or when
+distribution of work across threads is managed separately.
+
+*See also*: [`DataFrames.singlethreaded`](@ref) to disable multithreading only
+for a specific code block
+
+!!! note
+
+    This function is considered as experimental and may change or be removed once
+    a cross-package mechanism for multithreading configuration is developed.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(x=repeat(1:5, inner=2), y=1:10);
+
+julia> gd = groupby(df, :x);
+
+julia> counter = 0;
+
+julia> f(x) = (sleep(0.1); global counter += 1); # Thread-unsafe function
+
+julia> DataFrames.setmultithreading(false);
+
+julia> combine(gd, :y => f)
+5×2 DataFrame
+ Row │ x      y_f   
+     │ Int64  Int64 
+─────┼──────────────
+   1 │     1      1
+   2 │     2      2
+   3 │     3      3
+   4 │     4      4
+   5 │     5      5
+
+julia> DataFrames.setmultithreading(true);
+```
+"""
+function setmultithreading(enable::Bool)
+    if !enable && !SINGLETHREADING[]
+        Threads.atomic_add!(SINGLETHREADING_DEPTH, 1)
+    elseif enable && SINGLETHREADING[]
+        Threads.atomic_sub!(SINGLETHREADING_DEPTH, 1)
+    end
+    SINGLETHREADING[] = !enable
```

Comment:
```suggestion
    old_state = Threads.atomic_xchg!(SINGLETHREADING[], !enable)
    if !enable && !old_state
        Threads.atomic_add!(SINGLETHREADING_DEPTH, 1)
    elseif enable && old_state
        Threads.atomic_sub!(SINGLETHREADING_DEPTH, 1)
    end
```

Reply:
Your code seemed not thread safe. E.g. it could set the counter to negative value if `setmultithreadig(true)` were passed in multiple threads when `SINGLETHREADING[]` is `false` initially.